src/server/pfs/s3/object.go

```diff
@@ -0,0 +1,172 @@
+package s3
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+	"strings"
+	"fmt"
+	"crypto/md5"
+	"encoding/base64"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/pachyderm/pachyderm/src/client"
+)
+
+type objectHandler struct {
+	pc *client.APIClient
+}
+
+func newObjectHandler(pc *client.APIClient) *objectHandler {
+	return &objectHandler{pc: pc}
+}
+
+func (h *objectHandler) get(w http.ResponseWriter, r *http.Request) {
+	repo, branch, file := objectArgs(w, r)
+	branchInfo, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+	if branchInfo.Head == nil {
+		noSuchKeyError(w, r)
+		return
+	}
+	if isMeta(file) || strings.HasSuffix(file, "/") {
+		invalidFilePathError(w, r)
+		return
+	}
+
+	fileInfo, err := h.pc.InspectFile(branchInfo.Branch.Repo.Name, branchInfo.Head.ID, file)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+
+	timestamp, err := types.TimestampFromProto(fileInfo.Committed)
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	meta, err := getMeta(h.pc, branchInfo.Branch.Repo.Name, branchInfo.Head.ID, file)
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+	if meta != nil {
+		w.Header().Set("ETag", fmt.Sprintf("\"%s\"", meta.MD5))
+	}
+
+	reader, err := h.pc.GetFileReadSeeker(branchInfo.Branch.Repo.Name, branchInfo.Head.ID, file)
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	http.ServeContent(w, r, file, timestamp, reader)
+}
+
+func (h *objectHandler) put(w http.ResponseWriter, r *http.Request) {
+	repo, branch, file := objectArgs(w, r)
+	branchInfo, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+	if isMeta(file) || strings.HasSuffix(file, "/") {
+		invalidFilePathError(w, r)
+		return
+	}
+
+	expectedHash, ok := r.Header["Content-Md5"]
+	var expectedHashBytes []uint8
+	if ok && len(expectedHash) == 1 {
+		expectedHashBytes, err = base64.StdEncoding.DecodeString(expectedHash[0])
+		if err != nil || len(expectedHashBytes) != 16 {
+			invalidDigestError(w, r)
+			return
+		}
+	}
+
+	hasher := md5.New()
+	reader := io.TeeReader(r.Body, hasher)
+
+	client, err := h.pc.NewPutFileClient()
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	_, err = client.PutFileOverwrite(branchInfo.Branch.Repo.Name, branchInfo.Branch.Name, file, reader, 0)
+	if err != nil {
+		internalError(w, r, err)
+		if err = client.Close(); err != nil {
+			requestLogger(r).Errorf("could not close put file client: %v", err)
+		}
+		return
+	}
+
+	actualHashBytes := hasher.Sum(nil)
+	actualHash := fmt.Sprintf("%x", actualHashBytes)
+	failed := false
+	if expectedHashBytes != nil && !bytes.Equal(expectedHashBytes, actualHashBytes) {
+		badDigestError(w, r)
+		failed = true
+	}
+
+	if !failed {
+		meta := ObjectMeta { MD5: actualHash }
+		if err = putMeta(client, branchInfo.Branch.Repo.Name, branchInfo.Branch.Name, file, &meta); err != nil {
+			internalError(w, r, err)
+			failed = true
+		}
```

Comment:
The meta file is the only part of this PR that I'm architecturally not on board with. It seems like it's going to interact weirdly with other stuff, we filter the files from view in the `s3` handlers, but they'll still show in `list-file` and the like. Furthermore it seems like it only allows us to have md5s for objects that were put through the s3 api, which means in a lot of cases it won't be there. I think we should just use the file's `Hash` field as the `ETag`, the s3 spec doesn't guarantee that the `ETag` will be an MD5, and in fact in the case of multipart uploads it never will be, which is a pretty common case.

Reply:
That makes sense, but I'll have to substantially cut down the number of conformance tests run, since they specifically check against md5's

Reply:
Removing meta files means only 200 conformance tests can be run (vs 275). I think removal still makes sense though because:

1. As you said, the s3 spec doesn't specify what algo is used, so arguably this is a bug in the conformance tests themselves
2. The overall architecture should not be driven by what gets conformance tests to pass anyways

That said, the number of conformance tests is pretty tiny at this point

Reply:
Yeah, I think it still makes sense to remove it. Do they check that the ETag is actually an MD5? Or do they use the MD5 as a check for uploads? I think we can still support the latter.

Reply:
When putting a file, they check that the response etag matches an expected md5 value

Reply:
I see, it seems like that won't work with s3 proper though given that it only returns MD5s in some cases.